### PR TITLE
Fix error when setting outplanting levels due to latest RME expecting density as vector

### LINKED ISSUE #28: `set_outplant_deployment!` errors past RME version 1.0.28 due to expecting a float for density where RME expects a vector
With RME versions >1.0.28, using `set_outplant_deployment!` returns the following error:
```julia 
ERROR: ArgumentError: Call to RME failed with message: Error : Supplied vector is of size 0 should be 6.
```
This is due to later versions of the RME expecting a vector input for the density of corals where each element is the density of corals per species, whereas earlier versions expected a float. The ReefModEngine.jl function still expects a float so there's a mismatch. The ReefModEngine.jl function should be changed to allow for both float and vector input types, with a note on version requirements.

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
Resolves #28

### CODE REVIEW

File: src/deployment.jl [hunk 1]
```diff
@@ -1,41 +1,57 @@
 """
-    deployment_area(n_corals::Int64, max_n_corals::Int64, density::Float64, target_areas::Vector{Float64})::Tuple{Float64,Float64}
+    version_check()::Vector{Int64}
+
+Check RME version
+"""
+function version_check()::Vector{Int64}
+    rme_ver = @RME version()::Cstring
+    return parse.(Int64, split(rme_ver, '.'))
+end
```
Comment (ConnectedSystems) [inline src/deployment.jl:9]: I have an issue with the function name in that it doesn't perform a "check".

To me, it should be `rme_version()` or similar as it simply returns the version number (albeit separated).
Simply `version_check` can be confusing. It's unclear at first glance: What does it mean? Version of the Julia package?

I'm also hesitant to accept it returning a vector of integers. Version numbers should be treated as IDs firstly.

I also dislike how this is being used. Remember that if you're hardcoding index values, then it's usually a smell telling you there's a better way.

In this case, I would prefer:

1. a named tuple be returned: `(major=1, minor=0, patch=28)` (following [SemVer](https://semver.org/))
2. the function name be changed to `rme_version_info()`

File: src/deployment.jl [hunk 2]
```diff
@@ -177,17 +178,9 @@ function set_outplant_deployment!(
...
-        first_year::Int64,
-        last_year::Int64,
-        year_step::Int64,
-        area_km2::Vector{Float64},
-    )::Nothing
+    set_outplant_deployment!(name::String, reefset::String, max_effort::Int64, first_year::Int64, last_year::Int64, year_step::Int64, area_km2::Vector{Float64})::Nothing
 
-Set outplanting deployments across a range of years, automatically determining the
+    Set outplanting deployments across a range of years, automatically determining the
 coral deployment density to maintain the set grid size.
```
Comment (ConnectedSystems) [inline src/deployment.jl:214]: No indent for main text in docstrings

File: src/deployment.jl [hunk 3]
```diff
@@ -1,41 +1,57 @@
...
+    max_n_corals::Int64,
+    density::Union{Float64,Vector{Float64}},
+    target_areas::Vector{Float64}
+)::Union{Tuple{Float64,Float64},Tuple{Float64,Vector{Float64}}}
     req_area = area_needed(max_n_corals, density)
 
     # Divide by half (i.e., `* 0.5`) as RME simulates two deployments per year
-    mod_density = (n_corals * 0.5) / (req_area / m2_TO_km2)
-    deployment_area_pct = min((req_area / sum(target_areas)) * 100.0, 100.0)
+    mod_density = (n_corals * 0.5) ./ (req_area ./ m2_TO_km2)
```
Comment (ConnectedSystems) [inline src/deployment.jl:37]: Is this line a duplication of `area_needed()`?

File: src/deployment.jl [hunk 4]
```diff
@@ -1,41 +1,57 @@
...
+    deployment_area_pct = min((sum(req_area) / sum(target_areas)) * 100.0, 100.0)
```
Comment (ConnectedSystems) [inline src/deployment.jl:38]: Calculate `sum(req_area)` once beforehand and replace the repeated `sum()` here and below please.

File: src/deployment.jl [hunk 5]
```diff
@@ -71,13 +89,19 @@ function deployment_area(max_n_corals::Int64, target_areas::Vector{Float64})::Tu
...
+        )
     end
 
     @info "Determined min. deployment density to be: $(density) / m²"
-
     deployment_area_pct = min((req_area / sum(target_areas)) * 100.0, 100.0)
 
+    # In RME versions higher than 1.0.28 density needs to be a vector with each element representing density per species
+    version_vec = version_check()
+    density = (version_vec[3] <= 28) && (version_vec[1] == 1) ? density : fill(density, 6)
```
Comment (ConnectedSystems) [inline src/deployment.jl:104]: Ignoring the hardcoded indexing here, think about what would happen if the RME version advances to v1.2.28

File: src/deployment.jl [hunk 6]
```diff
@@ -130,15 +156,30 @@ function set_outplant_deployment!(
     last_year::Int64,
     year_step::Int64,
     area_km2::Vector{Float64},
-    density::Float64
+    density::Vector{Float64}
```
Comment (ConnectedSystems) [inline src/deployment.jl:159]: Doesn't this break backward compatibility?

File: docs/src/getting_started.md [hunk 7]
```diff
@@ -211,9 +211,6 @@ reset_rme()  # Reset RME to clear any previous runs
 # Note: if the Julia runtime crashes, check that the specified data file location is correct
 @RME runCreate(name::Cstring, start_year::Cint, end_year::Cint, RCP_scen::Cstring, gcm::Cstring, reps::Cint)::Cint
 
-# Add 3 DHW enhancement to outplanted corals
-set_option("restoration_dhw_tolerance_outplants", 3)
-
```
Comment (ConnectedSystems) [inline docs/src/getting_started.md:216]: Could you clarify why the outplanting removed in the example?

The commit message for this change just says "fix docs"?

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]

**Comment by Rosejoycrocker** [comment]
I believe I've addressed your comments @ConnectedSystems

File: docs/src/getting_started.md (hunk 7 above)
Comment (ConnectedSystems) [reply docs/src/getting_started.md:216]: Are we dropping backward compatibility?

File: src/deployment.jl [hunk 8]
```diff
@@ -1,41 +1,59 @@
 """
-    deployment_area(n_corals::Int64, max_n_corals::Int64, density::Float64, target_areas::Vector{Float64})::Tuple{Float64,Float64}
+    rme_version_info()::@NamedTuple{major::Int64, minor::Int64, patch::Int64}
+
+Get RME version
+"""
+function rme_version_info()::@NamedTuple{major::Int64, minor::Int64, patch::Int64}
+    rme_ver = @RME version()::Cstring
+    rme_ver = parse.(Int64, split(rme_ver, '.'))
```
Comment (ConnectedSystems) [inline src/deployment.jl:8]: No change required here, just nothing this is type unstable (going from string to vector of ints).

File: src/deployment.jl [hunk 9]
```diff
@@ -1,41 +1,59 @@
...
+"""
+    deployment_area(n_corals::Int64, max_n_corals::Int64, density::Union{Float64, Vector{Float64}}, target_areas::Vector{Float64})::Union{Tuple{Float64,Float64},Tuple{Float64,Vector{Float64}}}
 
 Determine deployment area for the expected number of corals to be deployed.
 
 # Arguments
 - `n_corals` : Number of corals,
 - `max_n_corals` : Expected maximum deployment effort (total number of corals in intervention set)
-- `density` : Stocking density per m²
+- `density` : Stocking density per m². In RME versions higher than 1.0.28 density needs to be a vector
```
Comment (ConnectedSystems) [inline src/deployment.jl:20]:
```suggestion
- `density` : Stocking density per m². In RME versions higher than v1.0.28, density is expected to be a vector, witha value for each functional group represented.
```

File: src/deployment.jl [hunk 10]
```diff
@@ -1,41 +1,59 @@
...
+    with each element representing density per species
```
Comment (ConnectedSystems) [inline src/deployment.jl:21]:
```suggestion
- `density` : Stocking density per m². In RME versions higher than v1.0.28, density is expected to be a vector
    with each element representing the density per functional group
```

File: src/deployment.jl [hunk 11]
```diff
@@ -130,15 +164,30 @@ function set_outplant_deployment!(
...
+        reefset::Cstring,
+        first_year::Cint,
+        last_year::Cint,
+        year_step::Cint
+    )::Cint
     @RME ivSetOutplantAreaPct(name::Cstring, area_pct::Cdouble)::Cint
-    @RME ivSetOutplantCountPerM2(name::Cstring, mod_density::Cdouble)::Cint
+
+    rme_version = rme_version_info()
+    if !((rme_version.patch <= 28) && (rme_version.minor == 0) && (rme_version.major == 1))
```
Comment (ConnectedSystems) [inline src/deployment.jl:184]: Reorder to be in logical order please: major, minor, patch.

**Comment by Rosejoycrocker** [comment]
I've addressed your comments @ConnectedSystems, let me know if there are others

**Comment by ConnectedSystems** [comment]
@Rosejoycrocker there's still the question of what to do with the example currently in the docs.

It would be best to show how to do intervention runs with the most recent RME we have.

**Comment by Rosejoycrocker** [comment]
> @Rosejoycrocker there's still the question of what to do with the example currently in the docs.
> 
> It would be best to show how to do intervention runs with the most recent version of RME we have at the very least, with a note on how to do it with older versions.

I'll add the removed example as a comment with a note about the version

File: docs/src/getting_started.md [hunk 12]
```diff
@@ -211,6 +211,10 @@ reset_rme()  # Reset RME to clear any previous runs
 # Note: if the Julia runtime crashes, check that the specified data file location is correct
 @RME runCreate(name::Cstring, start_year::Cint, end_year::Cint, RCP_scen::Cstring, gcm::Cstring, reps::Cint)::Cint
 
+# Adding dhw tolerance was removed in v1.0.31
+# Add 3 DHW enhancement to outplanted corals
+# set_option("restoration_dhw_tolerance_outplants", 3)
+
```
Comment (ConnectedSystems) [inline docs/src/getting_started.md:217]: You need to demonstrate what to do with versions > v1.0.31

From the docs and matlab files it looks like they added an `rme_iv_set_outplant_heat_tolerance` function.

The relevant RME functions are:

`ivSetOutplantHeatToleranceMeanDhw` and `ivSetOutplantHeatToleranceSdDhw`

Both of which expect a vector of values for each functional group.

---

Actually, this back and forth has gone for too long. I'm going to accept and merge this as is and address the above later.

**Review by ConnectedSystems — APPROVED** [review]

**Pull request merged**